### ISSUE TITLE
MacOS volume mount for /var/lib/containerd does not work

### ISSUE THREAD

**Issue by desagar** [kubernetes-sigs/kind#1989]
I am trying to make a number of large Docker images downloaded to my KinD cluster survive beyond the life of the cluster, so that when re-creating the cluster, it doesn’t result in the images being re-downloaded.  On Linux, I was able to get this to work very well by adding an `extraMount` of a host directory onto the containerPath `/var/lib/containerd` on my Kind node. When the cluster is deleted and re-created, `containerd` sees the images as already being present and doesn’t try to pull them again.

On MacOS, however, this does not work. I get the following types of errors during kind create cluster with the same type of extraMount in the config file.

**Note:** this is from `journalctl -xeu kubelet` on the `kind-control-plane` container on kind `0.9.0`

```
kind-control-plane kubelet[235]: E1222 15:06:00.324113     235 pod_workers.go:191] Error syncing pod 7627bfee00355e58439591bdeb159e85 ("kube-apiserver-kind-control-plane_kube-system(7627bfee00355e58439591bdeb159e85)
"), skipping: failed to "CreatePodSandbox" for "kube-apiserver-kind-control-plane_kube-system(7627bfee00355e58439591bdeb159e85)" with CreatePodSandboxError: "CreatePodSandbox for pod \"kube-apiserver-kind-control-plane_kube-system(
7627bfee00355e58439591bdeb159e85)\" failed: rpc error: code = Unknown desc = failed to create containerd task: failed to mount rootfs component &{overlay overlay [index=off workdir=/var/lib/containerd/io.containerd.snapshotter.v1.o
verlayfs/snapshots/125/work upperdir=/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/125/fs lowerdir=/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/3/fs]}: invalid argument: unknown"
```

I was initially using kind `0.7.0`, where I also see errors like the following on stdout during `kind create cluster`
```
[WARNING ImagePull]: failed to pull image k8s.gcr.io/kube-apiserver:v1.17.0.....etc etc......mount callback failed on /var/lib/containerd/tmpmounts/containerd-mount515057636: failed call to UtimesNanoAt for /var/lib/containerd/tmpmounts/containerd-mount515057636/bin/dnsdomainname: no such file or directory: unknown"
```

My guess based on other posts in this channel and elsewhere is that these issues might be because of the host machine’s  filesystem somehow not playing well with `containerd` inside the Kind node (I’m on 10.15.7 aka MacOS Catalina). 

Is there any way to make volume mounts from MacOS to `/var/lib/containerd` work correctly?

**Comment by BenTheElder** [comment]
On Linux these are a simple bind mount internally, but on mac if you mount to the host it has to translate the syscalls from Linux to Darwin and move data in and out of the VM, it might be possible to make work by changing the snapshotter to native using containerd config patches, alternatively it sounds like you just want to persist this directory so you could bind mind to an absolute path on mac somewhere inside the docker for mac VM maybe.

Currently this guide has an example of patching containerd config:
https://kind.sigs.k8s.io/docs/user/local-registry/#create-a-cluster-and-registry

If you check the one kind ships you want to patch this setting https://github.com/kubernetes-sigs/kind/blob/e36bf401e3d9792956f216855712606b3069d41e/images/base/files/etc/containerd/config.toml#L22

I can test this at some point but I'm just checking in breifly while out for a bit 😅

**Comment by desagar** [comment]
Thanks @BenTheElder I did come across that yesterday and tried it by docker execing into the kind-control-plane container and updating /etc/containerd/config.toml to set snapshotter="native", and then restarting containerd using systemctl. 

When I do `ctr image pull k8s.gcr.io/kube-apiserver:v1.19.1`, I get the same error during image unpacking:
```
unpacking linux/amd64 sha256:53c469bf931d50688253bed4cd4a970f16949bd5e4540d06aaeecc52d3e0e6a4...
INFO[0036] apply failure, attempting cleanup             error="failed to extract layer sha256:79d541cda6cb9a0c0e4aaa62aaea1f85b6b56544b5ad25e1e3369525ec0bf670: mount callback failed on /var/lib/containerd/tmpmounts/containerd-mount465545460: failed call to UtimesNanoAt for /var/lib/containerd/tmpmounts/containerd-mount465545460/usr/share/zoneinfo/localtime: operation not permitted: unknown" key="extract-566395600-dK_N sha256:79d541cda6cb9a0c0e4aaa62aaea1f85b6b56544b5ad25e1e3369525ec0bf670"
ctr: failed to extract layer sha256:79d541cda6cb9a0c0e4aaa62aaea1f85b6b56544b5ad25e1e3369525ec0bf670: mount callback failed on /var/lib/containerd/tmpmounts/containerd-mount465545460: failed call to UtimesNanoAt for /var/lib/containerd/tmpmounts/containerd-mount465545460/usr/share/zoneinfo/localtime: operation not permitted: unknown
```

Also, I'm not quite sure what you mean by "you could bind mind to an absolute path on mac somewhere inside the docker for mac VM maybe." - by specifying an extraMount in the kind config, I thought that is exactly what kind was doing when starting the docker container. 

Thanks a lot for responding during your time off. I really appreciate it.

**Comment by BenTheElder** [comment]
That's too bad re: native snapshotter.

> Also, I'm not quite sure what you mean by "you could bind mind to an absolute path on mac somewhere inside the docker for mac VM maybe." - by specifying an extraMount in the kind config, I thought that is exactly what kind was doing when starting the docker container.

It depends on the mount path you specify and how your docker for mac is configured. Some mounts like your $HOME on the host are setup so the VM / docker for mac do some special "passthrough" between the host and the containers inside the VM. This is NOT a trivial bind mount since it crosses the VM boundary.

A mount with hostpath /foo on the other hand may instead be a simple bind mount to /foo inside the VM to the container.

If docker for mac sees that the hostpath is under one of the paths shared to the VM, it instead does the translation between macOS and docker and the data winds up on macOS. If you just want to persist between clusters we could instead use a path that won't be shared to macOS when on mac, since we're actually running under the docker for mac VM.

--- 

I think we want to offer a built in feature for persisting pulls between clusters but the best version of it remains to be seen. One option coming up is leveraging the newly improved mirror host functionality in containerd / OCI slec, but we're going to need something that implements it on the server side and some testing to see if it performs as hoped. I don't think this will happen soon.

**Comment by desagar** [comment]
Thanks for the ideas/suggestions. I think by paths shared to the VM, you are referring to the paths set up under "File Sharing" in Docker for Mac. I tried with a path that is outside that, and kind seems to mount it, but it behaved the same way. However, I was able to get the following workaround to work, and wanted to drop it here for anyone else who may try to do this - in summary, rather than mounting a host path directly, I mounted the mountpath of a Docker volume. Details are below.

I created a docker volume (e.g.)
`docker volume create containerd`

Next, I got the docker volume's mount path (`docker volume inspect containerd`)

Then in the kind config, I used the docker volume's mount path as the hostPath, and did a regular a bind mount (evidently this hostPath is on my host - I think it's on the docker for Mac VM, but it still works):

```
    extraMounts:
      - hostPath: /var/lib/docker/volumes/containerd/_data
        containerPath: /var/lib/containerd
```

After this, the kind cluster starts up fine and I can use it. Even if I delete and re-create the cluster with the same config setting, it still seems to think that the images for any pods I used on the prior cluster, are "already present on machine".

I'm going to close this issue since there is a workaround that I can use on MacOS.
